python/tritonserver/_c/tritonserver_pybind.cc

```diff
@@ -896,107 +871,120 @@ class PyInferenceRequest
-    // goes through the request release cycle. This is due to
-    // a 'keep_alive' relationship is built between 'PyInferenceRequest'
-    // and 'PyServer': a request is associated with a server and the server
-    // should be kept alive until all associated requests is properly released.
-    // And here we exploit the 'keep_alive' utility in PyBind to guarantee so.
-    // See PyServer::InferAsync on how this field is set to avoid potential
-    // circular inclusion.
-    std::shared_ptr<PyInferenceRequest> request;
+      if (allocator != nullptr) {
+        ThrowIfError(TRITONSERVER_ResponseAllocatorDelete(allocator));
+      }
+    };
   };
-
-
-  void SetReleaseCallback(ReleaseFn release, const py::object& user_object)
+  void SetResponseAllocator()
+  {
+    // The response allocator is shared among all instances of this class, so it
+    // needs to be initialized once and once only.
+    // TODO: Why 'numpy.ones(2**27)' gets stuck when the response allocator is
+    //       static initialized?
+    static std::once_flag allocator_init_once;
+    std::call_once(allocator_init_once, [this]() {
+      struct TRITONSERVER_ResponseAllocator* allocator_raw = nullptr;
+      ThrowIfError(TRITONSERVER_ResponseAllocatorNew(
+          &allocator_raw, ResponseAllocatorAllocFn, ResponseAllocatorReleaseFn,
+          nullptr /* start_fn */));
+      allocator_.reset(allocator_raw);
+      ThrowIfError(TRITONSERVER_ResponseAllocatorSetQueryFunction(
+          allocator_.get(), ResponseAllocatorQueryFn));
+      ThrowIfError(TRITONSERVER_ResponseAllocatorSetBufferAttributesFunction(
+          allocator_.get(), ResponseAllocatorBufferAttributesFn));
+    });
+  }
+  static TRITONSERVER_Error* ResponseAllocatorQueryFn(
+      struct TRITONSERVER_ResponseAllocator* allocator, void* userp,
+      const char* tensor_name, size_t* byte_size,
+      TRITONSERVER_MemoryType* memory_type, int64_t* memory_type_id)
   {
-    request_callback_resource_.reset(
-        new CallbackResource(release, user_object));
-    ThrowIfError(TRITONSERVER_InferenceRequestSetReleaseCallback(
-        triton_object_, PyTritonRequestReleaseCallback,
-        request_callback_resource_.get()));
+    // TODO: Support GPU memory.
+    *memory_type = TRITONSERVER_MEMORY_CPU;
+    *memory_type_id = 0;
+    return nullptr;
   }
-
-  static void PyTritonRequestReleaseCallback(
-      struct TRITONSERVER_InferenceRequest* request, const uint32_t flags,
-      void* userp)
+  static TRITONSERVER_Error* ResponseAllocatorBufferAttributesFn(
+      struct TRITONSERVER_ResponseAllocator* allocator, const char* tensor_name,
+      struct TRITONSERVER_BufferAttributes* buffer_attributes, void* userp,
+      void* buffer_userp)
   {
-    py::gil_scoped_acquire gil;
-    auto cr = reinterpret_cast<CallbackResource*>(userp);
-    cr->release_fn(cr->request, flags, cr->user_object);
-    delete cr;
+    ThrowIfError(TRITONSERVER_BufferAttributesSetMemoryType(
```

Comment:
do we want to throw if error or return the error?

Reply:
sure, we can return the error. [Return if error when setting buffer attributes](https://github.com/triton-inference-server/core/pull/426/commits/c534d3a8a9e2b594b5e584f7c37b151a381fdcbf)